[LBSE] Fix some masker issues

rwlbuis opened the pull request:
#### 1e7aa8e2316ac92eb922cc9dbe5c283be9538979
<pre>
[LBSE] Fix some masker issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=265755">https://bugs.webkit.org/show_bug.cgi?id=265755</a>

Reviewed by Nikolas Zimmermann.

Clients of masks were not properly added in ReferencedSVGResources::referencedSVGResourceIDs, add logic for that, now
repaint rect results for the mask-object-bounding-box* tests are more as expected than before.
Also the mask-object-bounding-box* tests used the wrong unit attribute, to be consistent with the equivalent
clip-path tests &apos;maskContentUnits&apos; should be used. Finally, some more tests pass now, so update TestExpectations.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/svg/repaint/mask-object-bounding-box-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink.html:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed.html:
* LayoutTests/svg/repaint/mask-object-bounding-box.html:
* LayoutTests/svg/repaint/mask-user-space-on-use-expected.txt:
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedSVGResourceIDs):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::updateReferencedSVGResources):

Canonical link: <a href="https://commits.webkit.org/271631@main">https://commits.webkit.org/271631@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7ec60bcb2ef9dbdc39be4826718c7a9c72c143d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7759 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31691 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29685 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9884 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/5065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29361 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5561 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/25967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33029 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/5065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/7330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/25763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6937 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/ReferencedSVGResources.cpp:
```diff
@@ -114,6 +114,18 @@ Vector<std::pair<AtomString, QualifiedName>> ReferencedSVGResources::referencedS
         }
     }
+
+    if (style.hasPositionedMask()) {
```
Is it safe to do even for the legacy engine? Or do we perform extra work now? If so, you can add all of this behind LBSE guards

nikolaszimmermann replied:
Fixed by adding a guard, as Rob is on holidays.

nikolaszimmermann commented on Source/WebCore/rendering/ReferencedSVGResources.cpp:
```diff
@@ -114,6 +114,18 @@ Vector<std::pair<AtomString, QualifiedName>> ReferencedSVGResources::referencedS
+    if (style.hasPositionedMask()) {
+        // FIXME: We should support all the values in the CSS mask property, but for now just use the first mask-image if it's a reference.
+        auto* maskImage = style.maskImage();
+        auto reresolvedURL = maskImage ? maskImage->reresolvedURL(document) : URL();
```
Good catch r=me with one comment

nikolaszimmermann reviewed (approved):
r=me (fixed the one comment I had on my own, since Rob is on holidays).

rwlbuis committed:
[LBSE] Fix some masker issues
https://bugs.webkit.org/show_bug.cgi?id=265755

Reviewed by Nikolas Zimmermann.

Clients of masks were not properly added in ReferencedSVGResources::referencedSVGResourceIDs, add logic for that, now
repaint rect results for the mask-object-bounding-box* tests are more as expected than before.
Also the mask-object-bounding-box* tests used the wrong unit attribute, to be consistent with the equivalent
clip-path tests 'maskContentUnits' should be used. Finally, some more tests pass now, so update TestExpectations.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/svg/repaint/mask-object-bounding-box-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink.html:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed.html:
* LayoutTests/svg/repaint/mask-object-bounding-box.html:
* LayoutTests/svg/repaint/mask-user-space-on-use-expected.txt:
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedSVGResourceIDs):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::updateReferencedSVGResources):

Canonical link: https://commits.webkit.org/271631@main
